qualtran/bloqs/optimization/k_xor_sat/kikuchi_adjacency_list.py

```diff
@@ -0,0 +1,361 @@
+    r"""Given $(S, T)$, compute $k$ such that $T$ is the $k$-th non-zero entry in row $S$.
+
+    If $f(S, k)$ denotes the $k$-th non-zero entry in row $S$,
+    then this bloq maps $(S, f'(S, k), z, b)$ to $(S, f'(S, k), z \oplus k, b \oplus )$.
+    where $s$ is the sparsity, and $f'(S, k)$ is by extending $f$
+    such that for all $k \ge s$, $f'(S, k) = k$.
+    Using $f'$ ensures the computation is reversible.
+    Note: we must use the same extension $f'$ for both oracles.
+
+    This algorithm is described by the following pseudo-code:
+    ```
+    def reverse(S, f_S_k) -> k:
+        nnz := 0 # counter
+        for j in range(\bar{m}):
+            T := S \Delta U_j
+            if |T| == l:
+                nnz := nnz + 1
+            if T == f_S_k:
+                k ^= nnz
+    ```
+
+    Args:
+        inst: the kXOR instance $\mathcal{I}$.
+        ell: Kikuchi parameter $\ell$.
+
+    Registers:
+        S: index register to store $S \in {[n] \choose \ell}$.
+        k: non-zero entry index register
+    """
+
+    inst: KXorInstance
+    ell: SymbolicInt
+
+    @property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [
+                Register('S', self.index_dtype, shape=(self.ell,)),
+                Register('k', self.index_dtype, shape=(self.ell,)),
+                Register('T', self.index_dtype, shape=(self.ell,)),
+                Register('flag', QBit()),
+            ]
+        )
+
+    @property
+    def index_dtype(self) -> QUInt:
+        return QUInt(self.inst.index_bitsize)
+
+    def adjoint(self) -> 'IndexOfNonZeroColumn':
+        return self
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> 'BloqCountDictT':
+        m = self.inst.num_unique_constraints
+        ell, k = self.ell, self.inst.k
+
+        counts_forward = Counter[Bloq]()
+
+        # compute symmetric differences for each constraint
+        counts_forward[SymmetricDifference(ell, k, ell, self.index_dtype)] += m
+
+        # counter
+        counts_forward[AddK(self.index_dtype, 1).controlled()] += m
+
+        # compare T to f_S_k each time
+        counts_forward[Equals(self.index_dtype)] += m
+
+        # when T is equal (and counter is updated in this iteration), we can copy the result
+        counts_forward[And()] += m
+        counts_forward[CNOT()] += m  # flip the final flag (flipped at most once)
+
+        ### all counts
+        counts = Counter[Bloq]()
+
+        # copy the value of nnz (when final flag = 1)
+        counts[Xor(self.index_dtype).controlled()] += m
+
+        # if nothing matched (final flag = 0), copy k and flip the flag bit
+        counts[Xor(self.index_dtype).controlled()] += 1
+        counts[Xor(QBit())] += 1
+
+        for bloq, nb in counts_forward.items():
+            # compute and uncompute all intermediate values.
+            counts[bloq] += nb
+            counts[bloq.adjoint()] += nb
+
+        return counts
+
+
+@frozen
+class KikuchiNonZeroIndex(Bloq):
```

Comment:
I understand that the context to understand what's happening in this Bloq becomes clearer if the user also reads some of the other bloq docstrings (eg: ones for kxor_instance and other kikuchi matrix bloqs). It'd be helpeful (here and elsewhere) if we (a) either include a brief overview of the problem setup and define the necessary terms OR (b) add a sentence the refers the reader to docstrings of other bloqs to understand the background so the explanation in the docstring is easier to follow. I've highlighted two examples below of things that are not immediately clear without additional context.